backend/picamera_backend.py

```diff
@@ -0,0 +1,98 @@
+    video_feed = 2
+    base = 3
+    data = 4
+    message = 5
+
+class PiCameraBackend(Backend):
+    """ Gets video and displays using the `picamera` library.
+
+        This backend will only work when the `picamera` library is available,
+        i.e. when running on a Raspberry Pi. """
+
+    def __init__(self, width: int, height: int, publish_recording_status_func: PublishFunc):
+        super().__init__(width, height, publish_recording_status_func)
+
+        self.pi_camera = PiCamera(resolution=(self.width, self.height))
+
+        self.prev_overlays: Dict[OverlayLayer, self.pi_camera.PiOverlayRenderer] = {}
+
+    def start_video(self) -> None:
+        # Start displaying video feed. Non blocking, but runs forever in seperate thread.
+        self.pi_camera.start_preview(fullscreen=False, window=(*PI_WINDOW_TOP_LEFT, self.width, self.height))
+
+    def update_picamera_overlay(self, canvas: Canvas, layer: OverlayLayer) -> None:
+        """ Adds the overlay to a PiCamera preview, and if the overlay was already added,
+            removes the old instance. """
+        overlay = self.pi_camera.add_overlay(canvas.img, format="rgba", size=(self.width, self.height))
+        overlay.layer = layer.value
+        overlay.fullscreen = False
+        overlay.window = (*PI_WINDOW_TOP_LEFT, self.width, self.height)
+
+        # Rather than creating and swapping out overlays, the proper way to do this would be with overlay.update()
+        # Unfortunately, due to a bug in PiCamera 1.13, this will spam us with errors (which don't matter, but still)
+        # https://github.com/waveform80/picamera/issues/320
+        # https://www.raspberrypi.org/forums/viewtopic.php?t=190120
+        if layer in self.prev_overlays:
+            self.pi_camera.remove_overlay(self.prev_overlays[layer])
+        self.prev_overlays[layer] = overlay
+
+    def on_base_canvas_updated(self, base_canvas: Canvas) -> None:
+        self.update_picamera_overlay(base_canvas, OverlayLayer.base)
+
+    def on_canvases_updated(self, data_canvas: Canvas, message_canvas: Canvas) -> None:
+        """ Picamera will retain the overlay images until updated, so we only need
+            to do this once per overlay update. """
+        self.update_picamera_overlay(data_canvas, OverlayLayer.data)
+        self.update_picamera_overlay(message_canvas, OverlayLayer.message)
+
+    def _on_loop(self) -> None:
+        # Nothing to do
+        pass
+
+    def stop_video(self) -> None:
+        self.pi_camera.stop_preview()
+        self.stop_recording()
+        self.pi_camera.close()
+
+    def _start_recording(self) -> None:
+        try:
+            self.pi_camera.start_recording(self.recording_output_file)
+            self.recording = True
+            self.recording_start_time = time()
+            self.pi_camera.wait_recording(0.1)
+            self.send_recording_status()
+        except Exception:
+            self.recording = self.pi_camera.recording
```

Comment:
From the notes: https://picamera.readthedocs.io/en/release-1.10/api_camera.html#picamera.camera.PiCamera.recording

Seems like this variable would return `True` if an exception does occur assuming that `stop_recording()` is not called when the exception is raised. Any reason this is here? I would have just set `self.recording` to `False`, maybe I'm missing something?

Reply:
I'm trying to remember, I think this was to account for the case when it is told to start recording when it is already recording. In this case, I don't think the recording is stopped, so `self.pi_camera.recording` should be kept true. You're right though that the docs imply that `self.pi_camera.recording` may still be true when a fatal exception is raised. There are better, more clear ways I could have handled this, so I'll do that haha